segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -0,0 +1,450 @@
+                // Step 1: Create segment metadata.
+                for (String systemSegment : systemSegments) {
+                    SegmentMetadata segmentMetadata = SegmentMetadata.builder()
+                            .name(systemSegment)
+                            .ownerEpoch(epoch)
+                            .maxRollinglength(segmentRollingPolicy.getMaxLength())
+                            .build();
+                    segmentMetadata.setActive(true).setOwnershipChanged(true);
+                    segmentMetadata.checkInvariants();
+                    txn.create(segmentMetadata);
+                    txn.markPinned(segmentMetadata);
+                }
+
+                // We only need to apply the final truncate offset.
+                val finalTruncateOffsets = new HashMap<String, Long>();
+                val finalFirstChunkStartsAtOffsets = new HashMap<String, Long>();
+                val chunkStartOffsets = new HashMap<String, Long>();
+
+                // Step 2: For each epoch, find the corresponding system journal files.
+                for (int epochToRecover = 0; epochToRecover < epoch; epochToRecover++) {
+                    // Start scan with file index 0.
+                    int fileIndexToRecover = 0;
+                    while (chunkStorage.exists(getSystemJournalFileName(containerId, epochToRecover, fileIndexToRecover))) {
+                        val systemLogName = getSystemJournalFileName(containerId, epochToRecover, fileIndexToRecover);
+                        val info = chunkStorage.getInfo(systemLogName);
+                        val h = chunkStorage.openRead(systemLogName);
+                        byte[] contents = new byte[Math.toIntExact(info.getLength())];
+                        long fromOffset = 0;
+                        int remaining = contents.length;
+                        while (remaining > 0) {
+                            int bytesRead = chunkStorage.read(h, fromOffset, remaining, contents, Math.toIntExact(fromOffset));
+                            remaining -= bytesRead;
+                            fromOffset += bytesRead;
+                        }
+                        BufferedReader reader = new BufferedReader(new InputStreamReader(new ByteArrayInputStream(contents)));
+                        String line;
+
+                        while ((line = reader.readLine()) != null) {
+                            // Handle only whole records.
+                            if (line.startsWith(START_TOKEN) && line.endsWith(END_TOKEN)) {
+                                String[] records = line.split(RECORD_SEPARATOR);
+                                for (String record : records) {
+                                    String[] parts = record.split(":");
+                                    if (ADD_RECORD.equals(parts[0]) && parts.length == 5) {
+                                        String segmentName = parts[1];
+                                        String oldChunkName = parts[2];
```

Comment:
Remind me what the naming conventions are (you can point me to where it is documented). I'm trying to understand why we need an old and a new name is this add record.

Reply:
Having both old and new names makes it independent of the naming conventions itself. (Eg. In non backward compatible mode we'll use UUID as chunk name)